core/o11y/src/io_tracer.rs

```diff
@@ -0,0 +1,320 @@
+#![cfg(feature = "io_trace")]
+//! `tracing` layer to collect the sequence of IO operations executed.
+//!
+//! This module contains a somewhat quirky way of collecting a detailed log of
+//! which IO operations have been executed. The `tracing` crate is abused for
+//! this, as it provides an easy way to globally emit such events and organize
+//! them in a hierarchy of spans.
+//!
+//! The goal is to have something to look at to understand what requests are
+//! sent to the DB and for what reason. Not necessarily timing of events.(Timing
+//! might be affected by the sheer amount of details that this instrumentation
+//! is capable of producing.) But rather something that makes it obvious, for
+//! example, how effective an existing caching layer is for a workload by
+//! showing when the cache was hit and when it was missed.
+//!
+//! Another, equally important, use case is replaying the output for further
+//! analysis. The estimator has a replay command that understands the output
+//! produced by the IO trace.
+
+use std::collections::HashMap;
+use std::io::Write;
+use tracing::{span, Subscriber};
+use tracing_appender::non_blocking::{NonBlocking, WorkerGuard};
+use tracing_subscriber::fmt::MakeWriter;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::Layer;
+
+/// Tracing layer that produces a record of IO operations.
+pub struct IoTraceLayer {
+    make_writer: NonBlocking,
+    _guard: WorkerGuard,
```

Comment:
I believe this misses the point of the guard somewhat. The guard here exists so that the writer can tell when the application is about to exit and to give it an opportunity to flush any internal buffers and whatnot.

If this guard ends up being stored inside the subscriber, this guard will never get dropped. If flushing at exit is not a concern, then you can drop this guard immediately after it is constructed.

Reply:
Oh I see, makes sense.

Reply:
Turns out, dropping the guard early silently shuts down the writer thread. And no output is produced.

https://docs.rs/tracing-appender/latest/src/tracing_appender/non_blocking.rs.html#270-290

So I guess I will keep it as it is. Assuming that on normal shutdown, the buffered output still gets flushed, I think I am okay with this behaviour.

Reply:
Drop glue for global statics do not get run no matter how the process terminates.

Reply:
I guess that means I should be keeping the guard properly then. :) Done in the latest commit. (after the merge)